### PR TITLE
[xabuild.exe] [tests] use Microsoft.Build.Locator NuGet

### PR DESCRIPTION
**Pull request by jonathanpeppers**
In `Xamarin.ProjectTools` and `xabuild.exe`, we had some code looking
for Visual Studio 2017 installations on Windows.

Unfortunately we weren't doing it "the right way" and were merely
checking a few directories within `%ProgramFiles(x86)%`.

Making use of the `Microsoft.Build.Locator` NuGet package, allows for
simpler (and more correct) code for finding instances of Visual Studio.
I took the approach of choosing the "newest" installation, which works
well on my machine because it chooses my 15.7 preview installation
instead of the 15.6 one. Note that this NuGet only works on Windows, so
we can only use it in the cases we had the existing Windows
platform-specific code.

To enable a way to "choose" an instance of Visual Studio, I also added
checks for the `%VSINSTALLDIR%` environment variable. This allows usage
of the *Developer Command Prompt for VS 2017* or just setting the
environment variable system-wide.

Other changes:
- Had to raise `TargetFrameworkVersion` to at least 4.6, so I chose
4.6.2 to match Xamarin.Android.Build.Tests
- The IDE made some other adjustments in `*.csproj` files, but they
appeared to be "good" formatting changes or corrections.
- Updated `setup-windows.exe` to also take advantage of
`Microsoft.Build.Locator`, so you can now use a plain elevated `cmd.exe`

### CODE REVIEW

**Comment by dellis1972** [comment]
We could try to pick up the EnvVar from the VS2017 Dev command prompt I
guess, and use that if present?

**Comment by jonathanpeppers** [comment]
Yeah I should add a check for `VSINSTALLDIR` in both of these places.

This is just like we are doing in `setup-windows`: https://github.com/xamarin/xamarin-android/blob/cb30a13b144786891997184f2fcc161191db6f1c/tools/setup-windows/SetupWindows.cs#L48

**Comment by jonathanpeppers** [comment]
Need to work out issue with license at: https://www.nuget.org/packages/Microsoft.Build.Locator/

Link to license here: http://go.microsoft.com/fwlink/?LinkId=329770

So currently plan is to change this to a submodule of the open source repo, which is MIT: https://github.com/Microsoft/MSBuildLocator

**Comment by jonpryor** [comment]
Please also add an `external/MSBuildLocator.tpnitems` file, along the lines of https://github.com/xamarin/xamarin-android/blob/master/external/Java.Interop.tpnitems:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup Condition=" '$(TpnIncludeExternalDependencies)' == 'True' ">
    <ThirdPartyNotice Include="Microsoft/MSBuildLocator">
      <LicenseFile>$(MSBuildThisFileDirectory)\MSBuildLocator\LICENSE</LicenseFile>
      <SourceUrl>https://github.com/Microsoft/MSBuildLocator/</SourceUrl>
    </ThirdPartyNotice>
</Project>
```

**Comment by jonpryor** [comment]
[It didn't build](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3038):

```
xamarin-android/external/MSBuildLocator/src/MSBuildLocator/Microsoft.Build.Locator.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
```

**Comment by jonathanpeppers** [comment]
@jonpryor it must have worked for me because I have .NET Core installed.

Is that something we can depend on getting installed on the bots? If not, maybe we should rethink this.

**Comment by jonpryor** [comment]
> If not, maybe we should rethink this.

We could try referencing the existing source within another assembly, instead of building the assembly as-is.

**Comment by jonathanpeppers** [comment]
👍 I'll try this.

I also looked at the logs, and it is `bin/Debug/bin/xabuild Xamarin.Android-Tests.sln` that is failing, so that means it works for plain `msbuild`.  There may be another bug in `xabuild.exe` that could be fixed.

**Comment by jonathanpeppers** [comment]
@jonpryor I started setting up a project to link in the required files, and noticed it had another NuGet dependency:
* https://www.nuget.org/packages/Microsoft.VisualStudio.Setup.Configuration.Interop/
* License at: https://www.visualstudio.com/license-terms/vs-setup-sdk/

Does this mean we shouldn't use `Microsoft.Build.Locator`?

**Comment by jonathanpeppers** [comment]
Microsoft.VisualStudio.Setup.Configuration.Interop appears to be a development-time dependency: https://github.com/xamarin/xamarin-android/pull/1603/files#diff-f7564fc1e4607be2324a096c2f1ba062R3

It's also not present in the build output in `bin/Debug/bin` or the OSS bundle.

If we think this is OK, only two further comments to address:
[ ] Move the project into `src`
[ ] Setup the `*.tpnitems`

File: tools/Microsoft.Build.Locator/Microsoft.Build.Locator.csproj [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>bin\Debug\</OutputPath>
```
Comment (jonpryor) [inline tools/Microsoft.Build.Locator/Microsoft.Build.Locator.csproj]: Should the output path in fact be `tools/Microsoft.Build.Locator/bin/Debug/Microsoft.Build.Locator.dll`? More "conventional" in this project would be to to place it into its anticipated output location:

```xml
<OutputPath>..\..\bin\Debug\bin</OutputPath>
```

as it'll mostly be used by `xabuild.exe` (built into `$(topdir)/bin/$(Configuration)/bin`) and the unit tests (which will need to copy elsewhere), so setting `$(OutputPath)` to `$(topdir)/bin/$(Configuration)/bin` makes sense.

File: tools/Microsoft.Build.Locator/Microsoft.Build.Locator.csproj [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE;FEATURE_VISUALSTUDIOSETUP</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="Microsoft.VisualStudio.Setup.Configuration.Interop, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
```
Comment (jonpryor) [inline tools/Microsoft.Build.Locator/Microsoft.Build.Locator.csproj]: I'm trying to read the MSBuildLocator source and getting quite confused. Is `Microsoft.VisualStudio.Setup.Configuration.Interop` actually needed, along with the `FEATURE_VISUALSTUDIOSETUP` symbol? If we didn't have this `@(Reference)` what would happen? If we didn't define `FEATURE_VISUALSTUDIOSETUP` what would happen?

Comment (jonathanpeppers) [reply tools/Microsoft.Build.Locator/Microsoft.Build.Locator.csproj]: It doesn't appear to build if I remove either the `FEATURE_VISUALSTUDIOSETUP` symbol or the `Microsoft.VisualStudio.Setup.Configuration.Interop` reference.

I am not sure what `FEATURE_VISUALSTUDIOSETUP` is, by googling I only find it in the `msbuild` source: https://github.com/Microsoft/msbuild/blob/master/src/Directory.BeforeCommon.targets#L104

The reference being "development-time" only, this `EmbedInteropTypes` metadata must be what makes the assembly work without the reference: https://github.com/xamarin/xamarin-android/pull/1603/files#diff-5f202698c483a3dc92183954eb4428bfR37

**Comment by jonathanpeppers** [comment]
We are going to revisit using `vswhere` instead, as there is no licensing to deal with.

**Pull request closed without merging**